zarr/storage.py

```diff
@@ -2211,50 +2211,84 @@ class ABSStore(MutableMapping):
     ----------
     container : string
         The name of the ABS container to use.
+        .. deprecated::
+           Use ``client`` instead.
     prefix : string
         Location of the "directory" to use as the root of the storage hierarchy
         within the container.
     account_name : string
         The Azure blob storage account name.
+        .. deprecated:: 2.8.3
+           Use ``client`` instead.
     account_key : string
         The Azure blob storage account access key.
+        .. deprecated:: 2.8.3
+           Use ``client`` instead.
     blob_service_kwargs : dictionary
         Extra arguments to be passed into the azure blob client, for e.g. when
         using the emulator, pass in blob_service_kwargs={'is_emulated': True}.
+        .. deprecated:: 2.8.3
+           Use ``client`` instead.
     dimension_separator : {'.', '/'}, optional
         Separator placed between the dimensions of a chunk.
+    client : azure.storage.blob.ContainerClient, optional
+        And ``azure.storage.blob.ContainerClient`` to connect with. See
+        `here <https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.containerclient?view=azure-python>`_  # noqa
+        for more.
+
+        .. versionadded:: 2.8.3
 
     Notes
     -----
-    In order to use this store, you must install the Microsoft Azure Storage SDK for Python.
+    In order to use this store, you must install the Microsoft Azure Storage SDK for Python,
+    ``azure-storage-blob>=12.5.0``.
     """
 
-    def __init__(self, container, prefix='', account_name=None, account_key=None,
-                 blob_service_kwargs=None, dimension_separator=None):
-        from azure.storage.blob import BlockBlobService
-        self.container = container
-        self.prefix = normalize_storage_path(prefix)
-        self.account_name = account_name
-        self.account_key = account_key
+    def __init__(self, container=None, prefix='', account_name=None, account_key=None,
+                 blob_service_kwargs=None, dimension_separator=None,
+                 client=None,
+                 ):
         self._dimension_separator = dimension_separator
-        if blob_service_kwargs is not None:
-            self.blob_service_kwargs = blob_service_kwargs
-        else:  # pragma: no cover
-            self.blob_service_kwargs = dict()
-        self.client = BlockBlobService(self.account_name, self.account_key,
-                                       **self.blob_service_kwargs)
-
-    # needed for pickling
```

Comment:
I believe that `ContainerClient` instances are typically pickleable.

```python
In [4]: cc = azure.storage.blob.ContainerClient("https://daymeteuwest.blob.core.windows.net", container_name="daymet-zarr")

In [6]: pickle.loads(pickle.dumps(cc))
Out[6]: <azure.storage.blob._container_client.ContainerClient at 0x7f5ce149ae80>
```

Reply:
If you see the tests [here](https://github.com/zarr-developers/zarr-python/actions/runs/875115873) (even the ones that were cancelled) you'll see that the pickling has worked, but only for python >=3.7. For python <3.7 the pickling did not work. `TypeError: can't pickle _thread.RLock objects`. A little googling around seems to suggest that it will not work for python 3.6. So not sure what's the best way around this.

Reply:
Ah, thanks for that. We might be able to work around this by pickling the pieces of information needed to recreate the client, but I don't know if that's worth the maintenance effort since it's fixed in 3.7 and newer.

Reply:
https://github.com/zarr-developers/zarr-python/pull/759/commits/5a3307cac910eb981b99d3ca5f8c12ea38b03f77 and https://github.com/zarr-developers/zarr-python/pull/759/commits/a1e0d0c67d2d716255cdb5c07d852f27e4570f4d just skip the tests for 3.6 and earlier. If desired, I can implement the compatibility code for py36, with the caveat that it won't be 100% robust (but likely not worse than before).